core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -502,6 +503,9 @@ public boolean equals(Object o) {
         if (relocatingNodeId != null ? !relocatingNodeId.equals(that.relocatingNodeId) : that.relocatingNodeId != null) {
             return false;
         }
+        if (allocationId != null ? !allocationId.equals(that.allocationId) : that.allocationId != null) {
+            return false;
+        }
```

Comment:
I went through all equals I could find and ShardRouting as a key type in map. Didn't spot only problem. The only shady place is IndicesClusterService, but that's up next for cleaning.  I also think it's not a big risk as the allocation id is changed only on state changes , so other fields in ShardRouting are changed as well.